src/fd.cpp

```diff
@@ -0,0 +1,264 @@
+#ifdef _WIN32
+#ifndef FD_SETSIZE
+#define FD_SETSIZE 2048  // Increase max fds - only for select() fallback
+#endif
+#include <winsock2.h>
+#else
+#include <poll.h>
+#endif
+#include <Rcpp.h>
+#include <unistd.h>
+#include <cstdlib>
+#include "later.h"
+#include "callback_registry_table.h"
+
+#if R_VERSION < R_Version(4, 5, 0)
+
+inline SEXP R_mkClosure(SEXP formals, SEXP body, SEXP env) {
+  SEXP fun = Rf_allocSExp(CLOSXP);
+  SET_FORMALS(fun, formals);
+  SET_BODY(fun, body);
+  SET_CLOENV(fun, env);
+  return fun;
+}
+
+#endif
+
+pthread_attr_t pt_attr;
+int pt_attr_created = 0;
+
+extern CallbackRegistryTable callbackRegistryTable;
+extern SEXP later_fdcancel;
+
+typedef struct ThreadArgs_s {
+  SEXP callback;
+  double timeout;
+  std::unique_ptr<std::vector<int>> fds;
+  int rfds, wfds, efds;
+  int num_fds;
+  int loop;
+} ThreadArgs;
+
+static void thread_finalizer(SEXP xptr) {
+
+  if (R_ExternalPtrAddr(xptr) == NULL) return;
+#ifdef _WIN32
+  HANDLE *xp = (HANDLE *) R_ExternalPtrAddr(xptr);
+  CloseHandle(*xp);
+#else
+  pthread_t *xp = (pthread_t *) R_ExternalPtrAddr(xptr);
+#endif
+  R_Free(xp);
+
+}
+
+static void later_callback(void *arg) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+  Rf_eval(args->callback, R_GlobalEnv);
+  R_ReleaseObject(args->callback);
+
+}
+
+// CONSIDER: if necessary to add method for HANDLES on Windows. Would be different code to SOCKETs.
+// TODO: implement re-usable background thread.
+static void *select_thread(void *arg) {
+
+  std::unique_ptr<std::shared_ptr<ThreadArgs>> argsptr(static_cast<std::shared_ptr<ThreadArgs>*>(arg));
+  std::shared_ptr<ThreadArgs> args = *argsptr;
+
+  int result;
+  int *values;
+
+#ifndef POLLIN // fall back to select() for R on older Windows
```

Comment:
Correct me if I'm wrong but this would only cover Windows XP(!!) and earlier? I don't think we care about XP. That said, if it doesn't add any complexity (versus `#ifndef _WIN32` I guess?) I also don't mind you leaving it like this.

Reply:
Oh, I misread. If we don't care about XP then you could remove all the select() code, no?

Reply:
Yes, documentation says it's for Windows XP, but I got the errors for R4.1 and earlier e.g. here: https://github.com/shikokuchuo/later/actions/runs/11485049258/job/31964242382#step:3:2471
I can also re-produce on a local install of R4.1 on Windows. Must be something to do with the pre-UCRT toolchain. I'm ok to keep this code in for now - it's pretty much separate anyway.